languages/source.nim

```diff
@@ -941,6 +969,56 @@ const lang* = language:
         opp(bool_1, align(n_1, z_1, exp)) - opp(bool_2, align(n_2, z_2, exp)),
         exp, false)
 
+  func floatMul(a, b: float) -> float =
+    ## IEEE-754.2008 binary64 multiplication (with simplified nans).
+    case (a, b)
+    of Nan, _: Nan
+    of _, Nan: Nan
+    of Inf(bool), Zero(bool): Nan
+    of Zero(bool), Inf(bool): Nan
+    of Inf(bool_1), Inf(bool_2):
+      Inf(bool_1 xor bool_2)
+    of Inf(bool_1), Finite(bool_2, n, z):
+      Inf(bool_1 xor bool_2)
+    of Finite(bool_2, n, z), Inf(bool_1):
+      Inf(bool_1 xor bool_2)
+    of Finite(bool_1, n, z), Zero(bool_2):
+      Zero(bool_1 xor bool_2)
+    of Zero(bool_1), Finite(bool_2, n, z):
+      Zero(bool_1 xor bool_2)
+    of Zero(bool_1), Zero(bool_2):
+      Zero(bool_1 xor bool_2)
+    of Finite(bool_1, n_1, z_1), Finite(bool_2, n_2, z_2):
+      # x*(2^e_1) * y*(2^e_2) = x*y*(2^(e_1 + e_2))
+      # no need to go through `normalize`, because we know that the signficand
+      # doesn't need to / cannot be left-shifted
+      binaryRoundAux(bool_1 xor bool_2, n_1 * n_2, z_1 + z_2)
+
+  func floatDiv(a, b: float) -> float =
+    ## IEEE-754.2008 binary64 division (with simplified nans).
+    case (a, b)
+    of Nan, _: Nan
+    of _, Nan: Nan
+    of Inf(bool), Inf(bool):
+      Nan
+    of Zero(bool), Zero(bool):
+      Nan
+    of Inf(bool_1), Finite(bool_2, n, z):
+      Inf(bool_1 xor bool_2)
+    of Finite(bool_1, n, z), Inf(bool_2):
+      Zero(bool_1 xor bool_2)
+    of Inf(bool_1), Zero(bool_2):
+      Inf(bool_1 xor bool_2)
```

Comment:
TIL: that this is because zero in most cases is due to underflows, due to the nature of the spec, while inf is mostly overflows, so dividing by zero is seen as an overflow/underflow, that's also why we take the zero's sign into account.

crazy